From: Xi Ruoyao <xry111@mengyan1223.wang>
Date: Thu, 12 Aug 2021 20:31:59 +0000 (+0000)
Subject: mips: align stack in clone [BZ #28223]
X-Git-Url: https://sourceware.org/git/?p=glibc.git;a=commitdiff_plain;h=1f51cd9a860ee45eee8a56fb2ba925267a2a7bfe;hp=4cc79c217744743077bf7a0ec5e0a4318f1e6641

mips: align stack in clone [BZ #28223]

The MIPS O32 ABI requires 4 byte aligned stack, and the MIPS N64 and N32
ABI require 8 byte aligned stack.  Previously if the caller passed an
unaligned stack to clone the the child misbehaved.

Fixes bug 28223.
---

From: Xi Ruoyao <xry111@mengyan1223.wang>
Date: Fri, 13 Aug 2021 16:01:14 +0000 (+0000)
Subject: mips: increase stack alignment in clone to match the ABI
X-Git-Url: https://sourceware.org/git/?p=glibc.git;a=commitdiff_plain;h=0f62fe053273ff6c62ac95c59b7687c964737b00;hp=1f51cd9a860ee45eee8a56fb2ba925267a2a7bfe

mips: increase stack alignment in clone to match the ABI

In "mips: align stack in clone [BZ #28223]"
(commit 1f51cd9a860ee45eee8a56fb2ba925267a2a7bfe) I made a mistake: I
misbelieved one "word" was 2-byte and "doubleword" should be 4-byte.
But in MIPS ABI one "word" is defined 32-bit (4-byte), so "doubleword" is
8-byte [1], and "quadword" is 16-byte [2].

[1]: "System V Application Binary Interface: MIPS(R) RISC Processor
      Supplement, 3rd edition", page 3-31
[2]: "MIPSpro(TM) 64-Bit Porting and Transition Guide", page 23
---


Index: git/sysdeps/unix/sysv/linux/mips/clone.S
===================================================================
--- git.orig/sysdeps/unix/sysv/linux/mips/clone.S
+++ git/sysdeps/unix/sysv/linux/mips/clone.S
@@ -56,6 +56,15 @@ NESTED(__clone,4*SZREG,sp)
 #endif
 
 
+/* Align stack to 8/16 bytes per the ABI.  */
+#if _MIPS_SIM == _ABIO32
+ 	li		t0,-8
+#else
+	li		t0,-16
+#endif
+ 	and		a1,a1,t0
+
+
 	/* Sanity check arguments.  */
 	li		v0,EINVAL
 	beqz		a0,L(error)	/* No NULL function pointers.  */
